objax/jaxboard.py

```diff
@@ -95,27 +96,36 @@ def __call__(self):
 class SummaryWriter:
     """Writes entries to event files in the logdir to be consumed by Tensorboard."""
 
-    def __init__(self, logdir: str, queue_size: int = 5, write_interval: int = 5):
+    def __init__(self, logdir: str, queue_size: int = 5, write_interval: int = 5, only_first_host_writes: bool = True):
```

Comment:
Again, this multi-host setup trickles down every module and pollutes them. I think it should be handled at a higher level if possible.

Reply:
I think there are few options how to handle multi-host setup:

A. check `jax.host_id() == 0` in the training loop. This won't pollute `jaxboard.py` and `checkpoint.py`, however this would require user to always modify training loop

B. checkpoint and jaxboard perform writes only when `jax.host_id() == 0`.

C. checkpoint and jaxboard BY DEFAULT perform writes only when `jax.host_id() == 0`. However this behavior could be changed/altered. <-- this was original code of the pull request


Between options B and C, option B actually covers 99.99999% of cases, but theoretically I could imagine some weird situation when user may want to manually control which host perform writes (some weird model parallelism setup where several hosts actually perform writes, but they write different things).

Between option A and BC:
* **Advantage of options B and C**: any single-host code will work on multi-host setup pretty much without modifications. On the other hand, option A would require users to manually add `jax.host_id() == 0` on each IO op.
    * Theoretically it's better to add parallel barrier to the end of the multi-host program. But in practice code will work without parallel barrier, just last training step may be corrupted.
* **Advantage of option A**: it keeps the code of jaxboard and checkpoint cleaner.


WDYT?

Reply:
In light of your comments I prefer A.